### ISSUE TITLE
gradient_function/gradient_tapes with device annotations

### ISSUE THREAD

**Issue by WingsOfPanda** [tensorflow/tensorflow#54460]
**System information**
- TensorFlow version (you are using): 2.8
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Feature request: adding `colocate_gradients_with_ops` option to `tf.gradients` in 1.x to `tf.GradientTape` in 2.x.

This feature request/issue was 1st mentioned in https://github.com/tensorflow/tensorflow/issues/33688 almost two years ago by @olesalscheider . Allow me to quickly review the op in which the current behavior (still hold in tf 2.8) is described.

> Currently GradientTape.gradient() is executed on the device of the scope it is called in. Have a look at the following code:
> 
```
with tf.GradientTape() as tape:
    with tf.device('/gpu:1'):
        x = f1(input)
    with tf.device('/gpu:2'):
        x = f2(x)
    with tf.device('/gpu:0'):
        g = tape.gradient(x, f_vars)
```
> Here all gradient calculations will be carried out by GPU:0 and all variables needed for the gradient calculation will also be allocated on GPU:0. This is a problem if these temporary variables are too large to fit into the VRAM of GPU:0.
> 
> Please provide a way to execute the backward functions on the device of the corresponding forward function and allocate temporary variables for gradient calculation there. This allows to split a large model and distribute it among as many GPUs as necessary.
> 

Also, @olesalscheider provided a pr https://github.com/tensorflow/tensorflow/commit/a64ff0f2cda9d4e35ea450d4e945009a90ddee9a to achieved such feature and it get merged at the beginning, but shortly get rollbacked due to certain performance issues. 

I opened another pr https://github.com/tensorflow/tensorflow/pull/54510/commits/c292044a64700ecc3fa419b8247dfb1b61cf3ce7 to fit the current master (tf2.8) and build and test it. It looks to me the gradient can now be split correctly according to the device annotation.

**Will this change the current api? How?**
It will allow tf.GradientTape to do gradient in the device of the corresponding forward function. Hence, it would be possible to split the training of large models into as many possible GPUs as necessary

**Who will benefit with this feature?**
Anyone who wants to train large models that do not fit into the VRAM of a single GPU.

**Any Other info.**

It looks to me that the original post has gone silence... So I raise the issue again here to draw more attentions. It is a very import feature for peoples working in large image segmentation tasks. Some times the input tensor is so large and the model can not even be fitted into a single A100 card. 

Moreover, I noticed that in the discussion flow in the original post that there are some other work around like split into different tapes. I tried but with no luck. all gradient calculation still be allocated on gpu:0, and moreover, splitting tape would be very different for model like u-net which has a lot of skip-connections.

Last but not the least, I know for a fact that there is the Mash-Tensorflow which propose to do such job, but a native tensorflow support would be also very useful for people work on large models.

**Comment by WingsOfPanda** [comment]
I also opened this pr: [https://github.com/tensorflow/tensorflow/pull/54510/files](url) for your consideration. Need help to refine it further...

**Comment by WingsOfPanda** [comment]
I close this issue as the pr https://github.com/tensorflow/tensorflow/pull/54510/commits/c292044a64700ecc3fa419b8247dfb1b61cf3ce7 did provide what I desired. Thx!
